app/src/components/main/withPagination/withPagination.jsx

```diff
@@ -0,0 +1,117 @@
+import {
+  DEFAULT_PAGE_SIZE,
+  DEFAULT_PAGINATION,
+  defaultPaginationSelector,
+  PAGE_KEY,
+  SIZE_KEY,
+} from 'controllers/pagination';
+import { connectRouter } from 'common/utils';
+import { getStorageItem, updateStorageItem } from 'common/utils/storageUtils';
+import { userIdSelector } from 'controllers/user';
+import styles from './withPagination.scss';
+
+const cx = classNames.bind(styles);
+
+const PAGE = 'Page';
+const PAGE_SIZE = 'PageSize';
+const normalizeValue = (value) => (value > 0 ? value : undefined);
+
+export const withPagination = ({
+  namespace,
+  pageSizeOptions,
+  paginationSelector = defaultPaginationSelector,
+  defaultPageSize = DEFAULT_PAGE_SIZE,
+} = {}) => (WrappedComponent) =>
+  connectRouter(
+    (query) => ({
+      page: query[PAGE_KEY] && normalizeValue(Number(query[PAGE_KEY])),
+      size: query[SIZE_KEY] && normalizeValue(Number(query[SIZE_KEY])),
+    }),
+    {
+      updatePagination: (page, size) => ({ [PAGE_KEY]: page, [SIZE_KEY]: size }),
+    },
+    {},
+  )(({ captions, updatePagination, page: queryPage, size: queryPageSize, ...props }) => {
+    const userId = useSelector(userIdSelector);
+    const { totalElements, totalPages } = useSelector(paginationSelector);
+    const [pageSize, setPageSize] = useState();
+    const [page, setPage] = useState();
+
+    const updateStorage = (value, propertyName) => {
+      if (namespace) {
+        updateStorageItem(`${userId}_settings`, {
+          [`${namespace}${propertyName}`]: value,
+        });
+      }
+    };
+
+    useLayoutEffect(() => {
+      const userSettings = getStorageItem(`${userId}_settings`) || {};
+
+      const hashPage =
+        queryPage ||
+        normalizeValue(Number(userSettings[`${namespace}${PAGE}`])) ||
+        DEFAULT_PAGINATION[PAGE_KEY];
+
+      const hashPageSize =
+        queryPageSize ||
+        normalizeValue(Number(userSettings[`${namespace}${PAGE_SIZE}`])) ||
+        defaultPageSize;
+
+      setPageSize(hashPageSize);
+      setPage(hashPage);
+      updatePagination(hashPage, hashPageSize);
+      updateStorage(hashPage, PAGE);
+      updateStorage(hashPageSize, PAGE_SIZE);
+    }, []);
+
+    const changePage = (newPage) => {
+      setPage(newPage);
+      updateStorage(newPage, PAGE);
+      updatePagination(newPage, pageSize);
+    };
+
+    const changePageSize = (newPageSize) => {
+      setPageSize(newPageSize);
+      updateStorage(newPageSize, PAGE_SIZE);
+      updatePagination(page, newPageSize);
+    };
+
+    return (
+      <div className={cx('with-pagination-container')}>
+        <WrappedComponent {...props} />
+        <div className={cx('with-pagination')}>
+          <Pagination
+            pageSize={pageSize}
+            activePage={page}
+            totalItems={totalElements}
+            totalPages={totalPages}
+            pageSizeOptions={pageSizeOptions}
+            changePage={changePage}
+            changePageSize={changePageSize}
+            captions={captions}
```

Comment:
I suppose in most cases the `captions` will be the same for all pages, can we set messages you defined as default values?

Reply:
Done.